src/order/grade.lean

```diff
@@ -0,0 +1,334 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Violeta Hernández Palacios, Grayson Burton, Vladimir Ivanov
+-/
+import data.nat.interval
+import data.nat.succ_pred
+import order.atoms
+
+/-!
+# Graded orders
+
+This file defines graded orders, also known as ranked orders.
+
+A graded order is an order with a bottom element in which every element has some finite "height",
+that corresponds to how many elements you need to get down by to reach `⊥`.
+
+## Main declarations
+
+* `grade_order`: Graded orders.
+* `grade`: The grade of an element.
+* `grade_max_order`: Graded orders with maximal elements. All maximal elements have the same grade.
+* `max_grade`: The maximum grade in a `grade_max_order`.
+* `order_embedding.grade`: The grade of an element in a linear order as an order embedding.
+
+## Implementation notes
+
+The standard approach is to define graded orders as the bounded orders whose flags (maximal chains)
+all have the same finite length (see Stanley p. 99).
```

Comment:
I don't believe it's true that this is "standard", Wikipedia doesn't even use this as a definition, and oftentimes infinite orders are useful to consider as graded - it's just that one author prefers this.